GitPullRequest.Services/AzureDevOpsRepository.cs

```diff
@@ -0,0 +1,59 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+using LibGit2Sharp;
+
+namespace GitPullRequest.Services
+{
+    internal class AzureDevOpsRepository : RemoteRepository
+    {
+        readonly GitService gitService;
+
+        public AzureDevOpsRepository(GitService gitService, IRepository repo, string remoteName)
+            : base(gitService, repo, remoteName)
+        {
+            this.gitService = gitService;
+        }
+
+        public override int FindPullRequestForCanonicalName(string canonicalName)
+        {
+            var match = Regex.Match(canonicalName, "^refs/pull/([0-9]+)/merge$");
+            if (match.Success && int.TryParse(match.Groups[1].Value, out int number))
+            {
+                return number;
+            }
+
+            return -1;
+        }
+
+        protected override IDictionary<string, string> GetReferences(IRepository repo, string remoteName)
+        {
+            // for Azure DevOps we need to fetch PR branches so we can explore their history and get the commit before the automatic merge commit that is done on the server
+            gitService.Fetch(repo, remoteName, "+refs/pull/*/merge:refs/remotes/origin/pull/*/merge");
```

Comment:
I think we'll want to put each set of PRs under its own remote name:
```suggestion
            gitService.Fetch(repo, remoteName, $"+refs/pull/*/merge:refs/remotes/{remoteName}/pull/*/merge");
```

We could maybe also put _everything_ under different namespace (e.g. `/pulls/`):
```
            gitService.Fetch(repo, remoteName, $"+refs/pull/*/merge:refs/pulls/{remoteName}/pull/*/merge");
```

This might prevent all pull requests being fetched whenever the user does a `git fetch`?